zookeeper-server/src/main/java/org/apache/zookeeper/server/quorum/QuorumPeer.java

```diff
@@ -265,30 +283,67 @@ private LearnerType getType(String s) throws ConfigException {
             }
         }
 
+        public QuorumServer(long id, InetSocketAddress addr, InetSocketAddress electionAddr, LearnerType type) {
+            this(id, addr, electionAddr, null, null, type);
+        }
+
+        public QuorumServer(long id, InetSocketAddress addr, InetSocketAddress electionAddr, InetSocketAddress clientAddr, InetSocketAddress secureClientAddr, LearnerType type) {
+            this.id = id;
+            if (addr != null) {
+                this.addr.addAddress(addr);
+            }
+            if (electionAddr != null) {
+                this.electionAddr.addAddress(electionAddr);
+            }
+            this.type = type;
+            this.clientAddr = clientAddr;
+            this.secureClientAddr = secureClientAddr;
+
+            setMyAddrs();
+        }
+
         private static final String wrongFormat =
             " does not have the form server_config or server_config;client_config"
             + " where server_config is the pipe separated list of host:port:port or host:port:port:type"
-            + " and client_config is port or host:port";
+            + " and client_config is host:clientPort;host:secureClientPort or clientPort or host:clientPort"
+            + " or ';secureClientPort' or ';host:secureClientPort'";
 
         public QuorumServer(long sid, String addressStr) throws ConfigException {
             this.id = sid;
             LearnerType newType = null;
             String[] serverClientParts = addressStr.split(";");
             String[] serverAddresses = serverClientParts[0].split("\\|");
 
-            if (serverClientParts.length == 2) {
+            if (serverClientParts.length >= 2 && !serverClientParts[1].isEmpty()) {
                 String[] clientParts = ConfigUtils.getHostAndPort(serverClientParts[1]);
                 if (clientParts.length > 2) {
                     throw new ConfigException(addressStr + wrongFormat);
                 }
 
                 // is client_config a host:port or just a port
-                hostname = (clientParts.length == 2) ? clientParts[0] : "0.0.0.0";
+                String clientHostName = (clientParts.length == 2) ? clientParts[0] : "0.0.0.0";
```

Comment:
since processing for clientPort and secureClientPort is similar, you can dedup.  

by default, <firstPart-related to quorum>;<client-port related>;<secure-client-related>

Reply:
> by default, ;;

Didn't quite get that. Can you please elaborate?

secureClientPort can be committed entirely (backwards compatible behavior).

However, if it is specified, we ned to call `configureSSLAuth()`.

Reply:
We probably can rewrite this slightly better, but it has already been accepted as is in the OSS patch.

We can diverge from that and do some dedup or keep it as is.

Reply:
for now, it is ok. may be you can do it in OSS as well but later.